druid/src/app.rs

```diff
@@ -54,6 +56,43 @@ pub struct WindowDesc<T> {
     pub id: WindowId,
 }
 
+/// A description of a window to be instantiated. The user has to guarantee that this
+/// represents a `WindowDesc<T>` where `T` is the users `AppState`.
+///
+/// This includes a function that can build the root widget, as well as other
+/// window properties such as the title.
+pub struct AppStateWindowDesc {
```

Comment:
Naming: I would call this `AnyWindowDesc`.

Philosophy:  I get that this is a problem that needs to be solved for this PR to work, but I also think it's worth thinking about on its own.

What we need:

- some kind of type erasure or 'implicit `T`' for certain types of objects, like window descriptions and menus
- not to confuse the user
- to clutter the namespace as little as possible

What's the best way to do this? I'm not totally sure.

If we could have some internal type that did what this type is doing, that would be great. One way this might work is if we had convenience methods for things like making a new window or setting a menu; for instance creating a new window might just use a method on `EventCtx`, and then the user never needs to create the command or actually interact with the type erased type?

It might also be that we require a public type (that appears in the signature of the selector, for instance) but we don't actually need the user to ever interact with it?

Reply:
I though about some kind of wrapper type but I don't think it is possible with the Rust type system.
What I wanted to make more obvious: `MenuDesc` and others **must** represent the application state.
Adding methods for this to `EventCtx` seems like a possible solution.
Doing so could even allow equipping `EventCtx` with the `TypeId`s of eg. `MenuDesc<AppState>` to check it right away and panic if the type is wrong. That way the back trace will show exactly where the mismatch happened.
But if we want to keep it as a `Command` I doubt we can get around some kind of specific wrapper type.

Reply:
the nice  thing is that on `EventCtx` the method could take a concrete `MenuDesc`.

Reply:
How so? I mean `EventCtx` does not know what the `AppState` is, so how is it supposed to be typed for it?
If this could be done statically, than that would be absolutely my favorite.

Reply:
mm you're right I was thinking that `EventCtx` was paramaterized over `T`, but even in that case it doesn't tell us what the root data is.